File: dec01/calibration2.mjs
// https://adventofcode.com/2018/day/1
// part 2

import { readFileSync } from 'fs';

const digits = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine'];
let values = {};
digits.forEach((digit, i) => values[digit] = i + 1);

function* range(start, end) {
    for (let i = start; i <= end; i++) yield i;
}
[...range(0, 9)].forEach(i => values["" + i] = i);

const input = readFileSync('./input.txt', 'utf8');
const inputArray = input.split('\n');

let code = 0;

const find_digits = line => {
    // zero-width positive lookahead
    const re = new RegExp(`(?=(\\d|${digits.join('|')}))`, 'g');
    const matches = Array.from(line.matchAll(re), x => x[1]);
    const vals = matches.map(n => values[n]);
    return vals[0] * 10 + vals.at(-1);
};

inputArray
    .filter(line => line.trim().length > 0)
    .forEach(line => code += find_digits(line));

console.log(code);

File: dec01/calibration2.py
# https://adventofcode.com/2023/day/1
# Part 2
import regex as re  # regex can match overlapping patterns

DIGITS = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
VALUES = {d: i + 1 for i, d in enumerate(DIGITS)} | {str(i): i for i in range(10)}
DIGIT = re.compile(r'(\d|%s)' % '|'.join(DIGITS))


def find_first_and_last_digit(line):
    tmp = [VALUES[v] for v in DIGIT.findall(line, overlapped=True)]
    return tmp[0] * 10 + tmp[-1]

code = 0
for i, line in enumerate(open('input.txt'), start=1):
    if not line.strip():
        continue
    digits = find_first_and_last_digit(line)
    code += digits

print(code)  # 53340
